exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java

```diff
@@ -512,122 +509,122 @@ private void updateEstMaxBatchSize(RecordBatch incoming) {
-      // In the future HashAggregate may also need to perform some actions conditionally
-      // in the outer try block.
-
-      assert ! handlingSpills || currentIndex < Integer.MAX_VALUE;
 
-      outside:
-      while (true) {
+    while (true) {
 
-        // This would be called only once - after actual data arrives on incoming
-        if ( schema == null && incoming.getRecordCount() > 0 ) {
-          this.schema = incoming.getSchema();
-          // Calculate the number of partitions based on actual incoming data
-          delayedSetup();
-        }
+      // This would be called only once - first time actual data arrives on incoming
+      if ( schema == null && incoming.getRecordCount() > 0 ) {
+        this.schema = incoming.getSchema();
+        currentBatchRecordCount = incoming.getRecordCount(); // initialize for first non empty batch
+        // Calculate the number of partitions based on actual incoming data
+        delayedSetup();
+      }
 
-        // loop through existing records, aggregating the values as necessary.
-        if (EXTRA_DEBUG_1) {
-          logger.debug("Starting outer loop of doWork()...");
+      //
+      //  loop through existing records in this batch, aggregating the values as necessary.
+      //
+      if (EXTRA_DEBUG_1) {
+        logger.debug("Starting outer loop of doWork()...");
+      }
+      for (; underlyingIndex < currentBatchRecordCount; incIndex()) {
+        if (EXTRA_DEBUG_2) {
+          logger.debug("Doing loop with values underlying {}, current {}", underlyingIndex, currentIndex);
         }
-        for (; underlyingIndex < incoming.getRecordCount(); incIndex()) {
-          if (EXTRA_DEBUG_2) {
-            logger.debug("Doing loop with values underlying {}, current {}", underlyingIndex, currentIndex);
-          }
-          checkGroupAndAggrValues(currentIndex);
-          // If adding a group discovered a memory pressure during 1st phase, then start
-          // outputing some partition to free memory.
-          if ( earlyOutput ) {
-            outputCurrentBatch();
-            incIndex(); // next time continue with the next incoming row
-            return AggOutcome.RETURN_OUTCOME;
-          }
+        checkGroupAndAggrValues(currentIndex);
+        // If adding a group discovered a memory pressure during 1st phase, then start
+        // outputing some partition downstream in order to free memory.
+        if ( earlyOutput ) {
+          outputCurrentBatch();
+          incIndex(); // next time continue with the next incoming row
+          return AggOutcome.RETURN_OUTCOME;
         }
+      }
+
+      if (EXTRA_DEBUG_1) {
+        logger.debug("Processed {} records", underlyingIndex);
+      }
 
-        if (EXTRA_DEBUG_1) {
-          logger.debug("Processed {} records", underlyingIndex);
+      // Cleanup the previous batch since we are done processing it.
+      for (VectorWrapper<?> v : incoming) {
```

Comment:
If this is original code, then clearly it does the right thing else we'd have seen problems by now. So, fine to leave as-is.